### ISSUE TITLE
bug: rbac proxy in controller-manager Pod cannot access kube apiservice

### ISSUE THREAD

**Issue by MikeSpreitzer** [kubestellar/kubestellar#2355]
### Describe the bug

@dumb0002 tells me that the RBAC proxy in the KubeStellar controller-manager Pod is not authorized to make the calls that it needs on the kube apiservers of the KubeFlex hosting cluster. In his case, the WDS role in question is played by that hosting cluster.

### Steps To Reproduce

@dumb0002 ?

### Expected Behavior

That RBAC proxy can do its job, which includes asking the kube apiservice whether a given incoming (to the controller-manager Pod) request is authorized.

### Want to contribute?

- [ ] I would like to work on this issue.

**Comment by francostellari** [comment]
@dumb0002 can you explain a bit further what is needed. I did not develop the yaml for the RBAX proxy so I'm not very familiar with it. I guess it would be helpful to have a failing procedure so that I can work on the yaml to make it pass.

**Comment by dumb0002** [comment]
@francostellari, what's need are the following rbac for the kube-rbac-proxy: https://github.com/dumb0002/kubestellar/blob/ks-monitoring/monitoring/configuration/kube-rbac-proxy-config.yaml. 

I have been testing this scenario with prometheus to scrape the ks ctl manager that is behind that rbac-proxy. I will check for an easier way to test it.

**Comment by francostellari** [comment]
@MikeSpreitzer the ClusterRole that @dumb0002 needs [above](https://github.com/dumb0002/kubestellar/blob/ks-monitoring/monitoring/configuration/kube-rbac-proxy-config.yaml) is already in the config/rbac folder: https://github.com/kubestellar/kubestellar/blob/main/config/rbac/auth_proxy_role.yaml. But the make chart/kustomize does not merge them in the template yaml... why?

**Comment by MikeSpreitzer** [comment]
I gave it a little bit of a functional test. I used the current version of this PR, commit bda95d7ab3bc55e2ece7e38fa5823e6fc23c7d46 . I ran `test/e2e/run-test.sh` to create a working system. Then I did the following.

```shell
mspreitz@mjs13 kubestellar % kubectl config use-context kind-kubeflex
Switched to context "kind-kubeflex".

mspreitz@mjs13 kubestellar % kubectl get -n wds1-system deploy
NAME                             READY   UP-TO-DATE   AVAILABLE   AGE
kube-apiserver                   1/1     1            1           6m42s
kube-controller-manager          1/1     1            1           6m42s
kubestellar-controller-manager   1/1     1            1           4m37s
transport-controller             1/1     1            1           4m10s

mspreitz@mjs13 kubestellar % kubectl port-forward -n wds1-system deploy/kubestellar-controller-manager 8443
Forwarding from 127.0.0.1:8443 -> 8443
Forwarding from [::1]:8443 -> 8443
```

Then in another shell I did the following and examined the file that was produced; it was good.

```shell
mspreitz@mjs13 ~ % kubectl get --raw https://localhost:8443/metrics > /tmp/metrics.txt
mspreitz@mjs13 ~ % open /tmp/metrics.txt 
```

**Comment by MikeSpreitzer** [comment]
I tested and was unable to reproduce the problem. I used the `main` branch. I was able to scrape /metrics from port 8443 of the kubestellar-controller-manager using `kubectl get --raw https://localhost:8443/metrics > /tmp/metrics.txt` (after port-forwarding that port). I also examined the log of the proxy container, and found no complaint about access.

[krp.log](https://github.com/user-attachments/files/16512064/krp.log)

**Comment by dumb0002** [comment]
@MikeSpreitzer, port forward skips Token reviews, isn't it? I am now to provide a pass/fail test and verify the PR submitted by @francostellari

**Comment by MikeSpreitzer** [comment]
Before verifying the fix, we need to verify that we have a problem.

Port-forwarding is something external to the containers. The authorization checking is done by the kube-rbac-proxy container making a call to the kube API service. Port-forwarding does not change the logic inside the container nor intercept HTTP[S] calls to the kube API service.

**Comment by MikeSpreitzer** [comment]
```shell
mspreitz@mjs13 kubestellar % kubectl get --raw https://localhost:8443/metrics > /tmp/metrics.txt
                           
mspreitz@mjs13 kubestellar % open /tmp/metrics.txt 
(contents are good)

mspreitz@mjs13 kubestellar % curl http://localhost:8443/metrics
Client sent an HTTP request to an HTTPS server.
```

**Comment by MikeSpreitzer** [comment]
```shell
mspreitz@mjs13 kubestellar % curl -k https://localhost:8443/metrics
Unauthorized
```

**Comment by MikeSpreitzer** [comment]
Port forwarding is not adding the client TLS credentials from my kubeconfig.

**Comment by dumb0002** [comment]
Yes, I observed this problem with the prometheus pod. I will try to develop a test to validate it. The issue is due to the token review done by kube-rbac-proxy and I believe that is missing from the test done using port-forwarding. Looking to provide a way to reproduce the issue that I observed.

**Comment by dumb0002** [comment]
Also, we definitely need the rbac fixes provided in @francostellari  PR as they are need for the normal behavior of any kube-rbac-proxy.

**Comment by dumb0002** [comment]
@MikeSpreitzer @francostellari,  I posted an example of how to reproduce the issue in the submitted PR: https://github.com/kubestellar/kubestellar/pull/2379

**Comment by MikeSpreitzer** [comment]
This is odd. I was also able to reproduce the problem, in a slightly different way. I used a busybox container.

I checked out `main`(commit 8cf87bfa556ace20bb8485fe18b4430c1472e9f7). I used `test/e2e/run-test.sh` to create a running system. Despite successfully scraping the kubestellar-controller-manager using port-forwarding as reported above, I then tried the following method --- which got the opposite result.

I created a Pod from the following YAML.

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: test3
spec:
  containers:
  - name: bb
    image: busybox
    command:
    - sleep
    - "864001"
```

I looked up the Pod's cluster IP address; it was 10.244.0.26.

I then used `kubectl exec -it test3 -- sh` to get a shell inside the Pod. I `cd /var/run/secrets/kubernetes.io/serviceaccount` and then the following.

```shell
/var/run/secrets/kubernetes.io/serviceaccount # wget --no-check-certificate --header "Authorization: Bearer $(cat token)" -O- https://10.244.0.26:8443/metrics
Connecting to 10.244.0.26:8443 (10.244.0.26:8443)
wget: server returned error: HTTP/1.1 401 Unauthorized
```

Then I looked at the log from the kube-rbac-proxy container in the kubestellar-controller-manager Pod. I found the following complaints about that failed GET.

```
E0806 20:14:26.075865       1 webhook.go:154] Failed to make webhook authenticator request: tokenreviews.authentication.k8s.io is forbidden: User "system:serviceaccount:wds1-system:kubestellar-controller-manager" cannot create resource "tokenreviews" in API group "authentication.k8s.io" at the cluster scope
E0806 20:14:26.075905       1 proxy.go:73] Unable to authenticate the request due to an error: tokenreviews.authentication.k8s.io is forbidden: User "system:serviceaccount:wds1-system:kubestellar-controller-manager" cannot create resource "tokenreviews" in API group "authentication.k8s.io" at the cluster scope
```

**Comment by dumb0002** [comment]
@MikeSpreitzer, the command that you used with port-forwarding above is using the credentials from your kubeconfig file:
```
kubectl get --raw https://localhost:8443/metrics
```
That's my hypothesis of why that works and it also seems to be skipping the token review step from the kube-rbac-proxy. FYI, I noticed that kube-rbac-proxy only performs the token review in the first http call.
